File: WibioOtpRequestHelper.py
import requests
import json
import os
import random

class WibioOtpRequestHelper:
    SERVER = 'https://otpsandbox.wibiocard.com'
    USERNAME = '***************************'
    PASSWORD = '***************************'

    def __init__(self):
        self.session = requests.Session()
        self.cookie_file_name = f"{random.randint(100000000,1999999999):013}.txt"
        self.csrf_access_token = None

    def __del__(self):
        self.session.close()
        if os.path.exists(self.cookie_file_name):
            os.remove(self.cookie_file_name)

    def init(self):
        try:
            url = f"{self.SERVER}/admin/login"
            headers = {
                'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/32.0.1700.107 Chrome/32.0.1700.107 Safari/537.36'
            }
            data = {
                'username': self.USERNAME,
                'password': self.PASSWORD
            }
            response = self.session.post(url, headers=headers, data=data, verify=False)
            response.raise_for_status()
            
            cookies = response.cookies.get_dict()
            self.csrf_access_token = cookies.get('csrf_access_token')
            if self.csrf_access_token:
                self.session.headers.update({'X-CSRF-TOKEN': self.csrf_access_token})

            obj_session = response.json()
            if obj_session.get('result', {}).get('status') and obj_session.get('result', {}).get('value'):
                return obj_session['result']['status']
            else:
                raise Exception("cURL authentication failed!")
        except Exception as e:
            return str(e)

    def admin_assign(self, serial, user, pin='0000'):
        url = f"{self.SERVER}/admin/assign"
        data = {
            'serial': serial,
            'user': user,
            'pin': pin
        }
        response = self.session.post(url, data=data, verify=False)
        if response.status_code != 200:
            return "cURL request error!"
        return response.json()

    def admin_check(self, serial):
        url = f"{self.SERVER}/admin/check_serial"
        params = {
            'serial': serial
        }
        response = self.session.get(url, params=params, verify=False)
        if response.status_code != 200:
            return "cURL request error!"
        return response.json()

    def admin_disable_user(self, user):
        url = f"{self.SERVER}/admin/disable"
        data = {
            'user': user
        }
        response = self.session.post(url, data=data, verify=False)
        if response.status_code != 200:
            return "cURL request error!"
        return response.json()

    def admin_disable_token(self, token):
        url = f"{self.SERVER}/admin/disable"
        data = {
            'serial': token
        }
        response = self.session.post(url, data=data, verify=False)
        if response.status_code != 200:
            return "cURL request error!"
        return response.json()

    def admin_enable_user(self, user):
        url = f"{self.SERVER}/admin/enable"
        data = {
            'user': user
        }
        response = self.session.post(url, data=data, verify=False)
        if response.status_code != 200:
            return "cURL request error!"
        return response.json()

    def admin_enable_token(self, token):
        url = f"{self.SERVER}/admin/enable"
        data = {
            'serial': token
        }
        response = self.session.post(url, data=data, verify=False)
        if response.status_code != 200:
            return "cURL request error!"
        return response.json()

    def admin_get_token_owner(self, serial):
        url = f"{self.SERVER}/admin/getTokenOwner"
        params = {
            'serial': serial
        }
        response = self.session.get(url, params=params, verify=False)
        if response.status_code != 200:
            return "cURL request error!"
        return response.json()

    def admin_remove_user(self, user):
        url = f"{self.SERVER}/admin/remove"
        data = {
            'user': user
        }
        response = self.session.post(url, data=data, verify=False)
        if response.status_code != 200:
            return "cURL request error!"
        return response.json()

    def admin_remove_token(self, token):
        url = f"{self.SERVER}/admin/remove"
        data = {
            'serial': token
        }
        response = self.session.post(url, data=data, verify=False)
        if response.status_code != 200:
            return "cURL request error!"
        return response.json()

    def admin_reset_user(self, user):
        url = f"{self.SERVER}/admin/reset"
        data = {
            'user': user
        }
        response = self.session.post(url, data=data, verify=False)
        if response.status_code != 200:
            return "cURL request error!"
        return response.json()

    def admin_reset_token(self, token):
        url = f"{self.SERVER}/admin/reset"
        data = {
            'serial': token
        }
        response = self.session.post(url, data=data, verify=False)
        if response.status_code != 200:
            return "cURL request error!"
        return response.json()

    def admin_resync_user(self, user, otp1, otp2):
        url = f"{self.SERVER}/admin/resync"
        data = {
            'user': user,
            'otp1': otp1,
            'otp2': otp2
        }
        response = self.session.post(url, data=data, verify=False)
        if response.status_code != 200:
            return "cURL request error!"
        return response.json()

    def admin_resync_token(self, token, otp1, otp2):
        url = f"{self.SERVER}/admin/resync"
        data = {
            'serial': token,
            'otp1': otp1,
            'otp2': otp2
        }
        response = self.session.post(url, data=data, verify=False)
        if response.status_code != 200:
            return "cURL request error!"
        return response.json()

    def admin_set_user(self, user, pin='0000', MaxFailCount=10, SyncWindow=0, OtpLen=8, CounterWindow=0, hashlib='sha1', timeWindow=0, timeStep=30, timeShift=10, countAuthSuccessMax=0, countAuthSuccess=0, countAuth=0, countAuthMax=0, validityPeriodStart='', validityPeriodEnd='', phone=''):
        url = f"{self.SERVER}/admin/set"
        data = {
            'user': user,
            'pin': pin,
            'MaxFailCount': MaxFailCount,
            'SyncWindow': SyncWindow,
            'OtpLen': OtpLen,
            'CounterWindow': CounterWindow,
            'hashlib': hashlib,
            'timeWindow': timeWindow,
            'timeStep': timeStep,
            'timeShift': timeShift,
            'countAuthSuccessMax': countAuthSuccessMax,
            'countAuthSuccess': countAuthSuccess,
            'countAuth': countAuth,
            'countAuthMax': countAuthMax,
            'validityPeriodStart': validityPeriodStart,
            'validityPeriodEnd': validityPeriodEnd,
            'phone': phone
        }
        response = self.session.post(url, data=data, verify=False)
        if response.status_code != 200:
            return "cURL request error!"
        return response.json()

    def admin_set_token(self, token, pin='0000', MaxFailCount=10, SyncWindow=0, OtpLen=8, CounterWindow=0, hashlib='sha1', timeWindow=0, timeStep=30, timeShift=10, countAuthSuccessMax=0, countAuthSuccess=0, countAuth=0, countAuthMax=0, validityPeriodStart='', validityPeriodEnd='', phone=''):
        url = f"{self.SERVER}/admin/set"
        data = {
            'serial': token,
            'pin': pin,
            'MaxFailCount': MaxFailCount,
            'SyncWindow': SyncWindow,
            'OtpLen': OtpLen,
            'CounterWindow': CounterWindow,
            'hashlib': hashlib,
            'timeWindow': timeWindow,
            'timeStep': timeStep,
            'timeShift': timeShift,
            'countAuthSuccessMax': countAuthSuccessMax,
            'countAuthSuccess': countAuthSuccess,
            'countAuth': countAuth,
            'countAuthMax': countAuthMax,
            'validityPeriodStart': validityPeriodStart,
            'validityPeriodEnd': validityPeriodEnd,
            'phone': phone
        }
        response = self.session.post(url, data=data, verify=False)
        if response.status_code != 200:
            return "cURL request error!"
        return response.json()

    def admin_show_all(self, viewrealm='', tokeninfo_format='json'):
        url = f"{self.SERVER}/admin/show"
        params = {
            'viewrealm': viewrealm,
            'tokeninfo_format': tokeninfo_format
        }
        response = self.session.get(url, params=params, verify=False)
        if response.status_code != 200:
            return "cURL request error!"
        return response.json()

    def admin_show_user(self, user, viewrealm='', tokeninfo_format='json'):
        url = f"{self.SERVER}/admin/show"
        params = {
            'user': user,
            'viewrealm': viewrealm,
            'tokeninfo_format': tokeninfo_format
        }
        response = self.session.get(url, params=params, verify=False)
        if response.status_code != 200:
            return "cURL request error!"
        return response.json()

    def admin_show_token(self, token, viewrealm='', tokeninfo_format='json'):
        url = f"{self.SERVER}/admin/show"
        params = {
            'serial': token,
            'viewrealm': viewrealm,
            'tokeninfo_format': tokeninfo_format
        }
        response = self.session.get(url, params=params, verify=False)
        if response.status_code != 200:
            return "cURL request error!"
        return response.json()

    def admin_token_realm(self, token, realms):
        url = f"{self.SERVER}/admin/tokenrealm"
        data = {
            'serial': token,
            'realms': realms
        }
        response = self.session.post(url, data=data, verify=False)
        if response.status_code != 200:
            return "cURL request error!"
        return response.json()

    def admin_unassign_token(self, token):
        url = f"{self.SERVER}/admin/unassign"
        data = {
            'serial': token
        }
        response = self.session.post(url, data=data, verify=False)
        if response.status_code != 200:
            return "cURL request error!"
        return response.json()

    def admin_user_list(self, searchexpr='', realm='', resConf='', page=1, rp=100):
        url = f"{self.SERVER}/admin/userlist"
        params = {
            'searchexpr': searchexpr,
            'realm': realm,
            'resConf': resConf,
            'page': page,
            'rp': rp
        }
        response = self.session.get(url, params=params, verify=False)
        if response.status_code != 200:
            return "cURL request error!"
        return response.json()

File: WibioOtpRequestHelper.php
<?php

namespace App\Helpers;

use http;
use Illuminate\Support\Facades\Storage;

class WibioOtpRequestHelper
{
    const SERVER = 'https://otpsandbox.wibiocard.com';
    const USERNAME = '***************************';
    const PASSWORD = '***************************';
    const REALM = '***************************';

    private $ch;
    private $cookieFileName;
    private $csrf_access_token;

    public function __construct() {
        $this->ch = curl_init();
        $this->cookieFileName = str_pad(rand(100000000,1999999999), 13, '0', STR_PAD_LEFT).'.txt';
    }

    public function __destruct() {
        $this->end();
    }

/*init()
        POST /admin/login
        This function is used to authenticate the admin user against the WibioOtp server and retrive the session id to continue wirking.
        Parameters
            username – (required) the username
            password – (required) the password
*/
    public function init()
    {
        try
        {
            if (!function_exists("curl_init"))
                throw new Exception("cURL extension is not installed!");
                curl_setopt($this->ch, CURLOPT_URL, self::SERVER."/admin/login");
                curl_setopt($this->ch, CURLOPT_USERAGENT,'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/32.0.1700.107 Chrome/32.0.1700.107 Safari/537.36');
                curl_setopt($this->ch, CURLOPT_POST, true);
                curl_setopt($this->ch, CURLOPT_POSTFIELDS, "username=".self::USERNAME."&password=".self::PASSWORD);
                curl_setopt($this->ch, CURLOPT_RETURNTRANSFER, true);
                curl_setopt($this->ch, CURLOPT_COOKIESESSION, true);
                curl_setopt($this->ch, CURLOPT_SSL_VERIFYPEER, false);
                curl_setopt($this->ch, CURLOPT_COOKIEJAR, Storage::disk('local')->path($this->cookieFileName));
                curl_setopt($this->ch, CURLOPT_COOKIEFILE, Storage::disk('local')->path($this->cookieFileName));

            $session = curl_exec($this->ch);
                if (curl_error($this->ch)) throw new Exception("cURL session not retrived!");

            $cookies = curl_getinfo($this->ch, CURLINFO_COOKIELIST);
            $getCookie = false;
            foreach (explode("\t", $cookies[1]) as $cookie) {
                if ($getCookie) $this->csrf_access_token = $cookie;
                if ($cookie == 'csrf_access_token') $getCookie = true;
            }
            curl_setopt($this->ch, CURLOPT_HTTPHEADER, array('X-CSRF-TOKEN: ' . $this->csrf_access_token));

            $ObjectSession = json_decode($session);
            if($ObjectSession && $ObjectSession->{'result'}->{'status'} == true && $ObjectSession->{'result'}->{'value'} == true)
            {
                return $ObjectSession->{'result'}->{'status'};
            }
            else
                throw new Exception("cURL authentication failed!");
        }
        catch (Exception $e)
        {
            return $e->getMessage();
        }
    }

/*end()
        This function is used to close the session with the WibioOtp server and delete the local cookie file.
*/
    public function end()
    {
        curl_close($this->ch);
        Storage::disk('local')->delete($this->cookieFileName);
    }

/*assign()
        POST /admin/assign
        assigns a token to a user, i.e. a binding between the token and the user is created.
        Parameters
        serial – (required) the serial number / identifier of the token
        user – (required) login user name
        pin – (optional) - the pin of the user pass
*/
    public function adminAssign($serial, $user, $pin='0000')
    {
        curl_setopt($this->ch, CURLOPT_URL, self::SERVER."/admin/assign");
        curl_setopt($this->ch, CURLOPT_POST, true);
        curl_setopt($this->ch, CURLOPT_POSTFIELDS, "serial=".$serial."&user=".$user."&pin=".$pin);
        curl_setopt($this->ch, CURLOPT_SSL_VERIFYPEER, false);
        $answer = curl_exec($this->ch);
        if (curl_error($this->ch)) return ("cURL request error!");
        return json_decode($answer);
    }

/*check_serial()
        GET /admin/check_serial
        This function checks, if a given serial will be unique. It returns True if the serial does not yet exist and new_serial as a new value for a serial, that does not exist, yet
        Parameters
            serial – (required) the serial number / identifier of the token
*/
    public function adminCheck($serial)
    {
        curl_setopt($this->ch, CURLOPT_URL, self::SERVER."/admin/check_serial?serial=".$serial);
        curl_setopt($this->ch, CURLOPT_POST, false);
        curl_setopt($this->ch, CURLOPT_SSL_VERIFYPEER, false);
        $answer = curl_exec($this->ch);
        if (curl_error($this->ch)) return ("cURL request error!");
        return json_decode($answer);
    }

/*disable()
        POST /admin/disable
        disables a token given by serial or all tokens of a user
        Parameters
            serial – the token serial
            user – the user for whom all tokens will be disabled
*/
    public function adminDisableUser($user)
    {
        curl_setopt($this->ch, CURLOPT_URL, self::SERVER."/admin/disable");
        curl_setopt($this->ch, CURLOPT_POST, true);
        curl_setopt($this->ch, CURLOPT_POSTFIELDS, "user=".$user);
        curl_setopt($this->ch, CURLOPT_SSL_VERIFYPEER, false);
        $answer = curl_exec($this->ch);
        if (curl_error($this->ch)) return ("cURL request error!");
        return json_decode($answer);
    }
    public function adminDisableToken($token)
    {
        curl_setopt($this->ch, CURLOPT_URL, self::SERVER."/admin/disable");
        curl_setopt($this->ch, CURLOPT_POST, true);
        curl_setopt($this->ch, CURLOPT_POSTFIELDS, "serial=".$token);
        curl_setopt($this->ch, CURLOPT_SSL_VERIFYPEER, false);
        $answer = curl_exec($this->ch);
        if (curl_error($this->ch)) return ("cURL request error!");
        return json_decode($answer);
    }

/*enable()
        POST /admin/enable
        enables a token or all tokens of a user
        Parameters
            serial – (optional), the token serial number
            user – (optional), will enable all tokens of a user
*/
    public function adminEnableUser($user)
    {
        curl_setopt($this->ch, CURLOPT_URL, self::SERVER."/admin/enable");
        curl_setopt($this->ch, CURLOPT_POST, true);
        curl_setopt($this->ch, CURLOPT_POSTFIELDS, "user=".$user);
        curl_setopt($this->ch, CURLOPT_SSL_VERIFYPEER, false);
        $answer = curl_exec($this->ch);
        if (curl_error($this->ch)) return ("cURL request error!");
        return json_decode($answer);
    }
    public function adminEnableToken($token)
    {
        curl_setopt($this->ch, CURLOPT_URL, self::SERVER."/admin/enable");
        curl_setopt($this->ch, CURLOPT_POST, true);
        curl_setopt($this->ch, CURLOPT_POSTFIELDS, "serial=".$token);
        curl_setopt($this->ch, CURLOPT_SSL_VERIFYPEER, false);
        $answer = curl_exec($this->ch);
        if (curl_error($this->ch)) return ("cURL request error!");
        return json_decode($answer);
    }

/*getTokenOwner()
        GET /admin/getTokenOwner
        provide the userinfo of the token, which is specified as serial
        Parameters
            serial – the serial number of the token
*/
    public function adminGetTokenOwner($serial)
    {
        curl_setopt($this->ch, CURLOPT_URL, self::SERVER."/admin/getTokenOwner?serial=".$serial);
        curl_setopt($this->ch, CURLOPT_POST, false);
        curl_setopt($this->ch, CURLOPT_SSL_VERIFYPEER, false);
        $answer = curl_exec($this->ch);
        if (curl_error($this->ch)) return ("cURL request error!");
        return json_decode($answer);
    }

/*initToken()
        POST /admin/init
        initializes a token
        Parameters
            otpkey – (required) the hmac Key of the token
            genkey – (required) =1, if key should be generated. We either need otpkey or genkey
            keysize – (optional) either 20 or 32. Default is 20
            serial – (required) the serial number / identifier of the token
            description – (optional)
            pin – (optional) the pin of the user pass
            user – (optional) login user name
            realm – (optional) realm of the user
            type – (opt:ional) the type of the token
            tokenrealm – (optional) the realm a token should be put into
            otplen – (optional) length of the OTP value
            hashlib – (optional) used hashlib sha1 oder sha256
        NB: ocra2 & qrtoken are supported
*/
    public function adminInitOtpToken($otpkey, $serial, $pin='0000', $user='', $realm='', $type='HMAC', $tokenrealm='', $otplen=8, $hashlib='sha1')
    {
        $genkey = ($otpkey == "") ? 1 : 0;
        curl_setopt($this->ch, CURLOPT_URL, self::SERVER."/admin/init");
        curl_setopt($this->ch, CURLOPT_POST, true);
        curl_setopt($this->ch, CURLOPT_POSTFIELDS, "otpkey=".$otpkey."&genkey=".$genkey."&serial=".$serial."&pin=".$pin."&user=".$user."&realm=".strtolower($realm)."&type=".$type."&tokenrealm=".$tokenrealm."&otplen=".$otplen."&hashlib=".$hashlib);
        curl_setopt($this->ch, CURLOPT_SSL_VERIFYPEER, false);
        $answer = curl_exec($this->ch);
        if (curl_error($this->ch)) return ("cURL request error!");
        return json_decode($answer);
    }

/*remove()
        POST /admin/remove
        deletes either a certain token given by serial or all tokens of a user
        Parameters
            serial –the serial number of the token
            user – (optional) , will delete all tokens of a user
*/
    public function adminRemoveUser($user)
    {
        curl_setopt($this->ch, CURLOPT_URL, self::SERVER."/admin/remove");
        curl_setopt($this->ch, CURLOPT_POST, true);
        curl_setopt($this->ch, CURLOPT_POSTFIELDS, "user=".$user);
        curl_setopt($this->ch, CURLOPT_SSL_VERIFYPEER, false);
        $answer = curl_exec($this->ch);
        if (curl_error($this->ch)) return ("cURL request error!");
        return json_decode($answer);
    }
    public function adminRemoveToken($token)
    {
        curl_setopt($this->ch, CURLOPT_URL, self::SERVER."/admin/remove");
        curl_setopt($this->ch, CURLOPT_POST, true);
        curl_setopt($this->ch, CURLOPT_POSTFIELDS, "serial=".$token);
        curl_setopt($this->ch, CURLOPT_SSL_VERIFYPEER, false);
        $answer = curl_exec($this->ch);
        if (curl_error($this->ch)) return ("cURL request error!");
        return json_decode($answer);
    }

/*reset()
        POST /admin/reset
        reset the FailCounter of a Token
        Parameters
            serial (user or) – to identify the tokens
*/
    public function adminResetUser($user)
    {
        curl_setopt($this->ch, CURLOPT_URL, self::SERVER."/admin/reset");
        curl_setopt($this->ch, CURLOPT_POST, true);
        curl_setopt($this->ch, CURLOPT_POSTFIELDS, "user=".$user);
        curl_setopt($this->ch, CURLOPT_SSL_VERIFYPEER, false);
        $answer = curl_exec($this->ch);
        if (curl_error($this->ch)) return ("cURL request error!");
        return json_decode($answer);
    }
    public function adminResetToken($token)
    {
        curl_setopt($this->ch, CURLOPT_URL, self::SERVER."/admin/reset");
        curl_setopt($this->ch, CURLOPT_POST, true);
        curl_setopt($this->ch, CURLOPT_POSTFIELDS, "serial=".$token);
        curl_setopt($this->ch, CURLOPT_SSL_VERIFYPEER, false);
        $answer = curl_exec($this->ch);
        if (curl_error($this->ch)) return ("cURL request error!");
        return json_decode($answer);
    }

/*resync()
    POST /admin/resync
    this function resync the token, if the counter on server side is out of sync with the physical token.
    Parameters
        serial – serial or user (required)
        user – s.o.
        otp1 – the next otp to be found
        otp2 – the next otp after the otp1
*/
    public function adminResyncUser($user, $otp1, $otp2)
    {
        curl_setopt($this->ch, CURLOPT_URL, self::SERVER."/admin/resync");
        curl_setopt($this->ch, CURLOPT_POST, true);
        curl_setopt($this->ch, CURLOPT_POSTFIELDS, "user=".$user."&otp1=".$otp1."&otp2=".$otp2);
        curl_setopt($this->ch, CURLOPT_SSL_VERIFYPEER, false);
        $answer = curl_exec($this->ch);
        if (curl_error($this->ch)) return ("cURL request error!");
        return json_decode($answer);
    }
    public function adminResyncToken($token, $otp1, $otp2)
    {
        curl_setopt($this->ch, CURLOPT_URL, self::SERVER."/admin/resync");
        curl_setopt($this->ch, CURLOPT_POST, true);
        curl_setopt($this->ch, CURLOPT_POSTFIELDS, "serial=".$token."&otp1=".$otp1."&otp2=".$otp2);
        curl_setopt($this->ch, CURLOPT_SSL_VERIFYPEER, false);
        $answer = curl_exec($this->ch);
        if (curl_error($this->ch)) return ("cURL request error!");
        return json_decode($answer);
    }

/*set()
        POST /admin/set
        this function is used to set many different values of a token.
        Parameters
            serial – (optional)
            user – (optional)
            pin – (optional) - set the OTP PIN
            MaxFailCount – (optional) - set the maximum fail counter of a token
            SyncWindow – (optional) - set the synchronization window of the token
            OtpLen – (optional) - set the OTP Lenght of the token
            CounterWindow – (optional) - set the counter window (blank presses)
            hashlib – (optional) - set the hashing algo for HMAC tokens. This can be sha1, sha256, sha512
            timeWindow – (optional) - set the synchronize window for timebased tokens (in seconds)
            timeStep – (optional) - set the timestep for timebased tokens (usually 30 or 60 seconds)
            timeShift – (optional) - set the shift or timedrift of this token
            countAuthSuccessMax – (optional) - set the maximum allowed successful authentications
            countAuthSuccess – (optional) - set the counter of the successful authentications
            countAuth – (optional) - set the counter of authentications
            countAuthMax – (optional) - set the maximum allowed authentication tries
            validityPeriodStart – (optional) - set the start date of the validity period. The token can not be used before this date
            validityPeriodEnd – (optional) - set the end date of the validaity period. The token can not be used after this date
            phone – set the phone number for an SMS token
*/
    public function adminSetUser($user, $pin='0000', $MaxFailCount=10, $SyncWindow=0, $OtpLen=8, $CounterWindow=0, $hashlib='sha1', $timeWindow=0, $timeStep=30, $timeShift=10, $countAuthSuccessMax=0, $countAuthSuccess=0, $countAuth=0, $countAuthMax=0, $validityPeriodStart=0, $validityPeriodEnd=0, $phone='')
    {
        curl_setopt($this->ch, CURLOPT_URL, self::SERVER."/admin/set");
        curl_setopt($this->ch, CURLOPT_POST, true);
        curl_setopt($this->ch, CURLOPT_POSTFIELDS, "user=".$user."&pin=".$pin."&MaxFailCount=".$MaxFailCount."&SyncWindow=".$SyncWindow."&OtpLen=".$OtpLen."&CounterWindow=".$CounterWindow."&hashlib=".$hashlib."&timeWindow=".$timeWindow."&timeStep=".$timeStep."&timeShift=".$timeShift."&countAuthSuccessMax=".$countAuthSuccessMax."&countAuthSuccess=".$countAuthSuccess."&countAuth=".$countAuth."&countAuthMax=".$countAuthMax."&validityPeriodStart=".$validityPeriodStart."&validityPeriodEnd=".$validityPeriodEnd."&phone=".$phone);
        curl_setopt($this->ch, CURLOPT_SSL_VERIFYPEER, false);
        $answer = curl_exec($this->ch);
        if (curl_error($this->ch)) return ("cURL request error!");
        return json_decode($answer);
    }
    public function adminSetToken($token, $pin='0000', $MaxFailCount=10, $SyncWindow=0, $OtpLen=8, $CounterWindow=0, $hashlib='sha1', $timeWindow=0, $timeStep=30, $timeShift=10, $countAuthSuccessMax=0, $countAuthSuccess=0, $countAuth=0, $countAuthMax=0, $validityPeriodStart=0, $validityPeriodEnd=0, $phone='')
    {
        curl_setopt($this->ch, CURLOPT_URL, self::SERVER."/admin/set");
        curl_setopt($this->ch, CURLOPT_POST, true);
        curl_setopt($this->ch, CURLOPT_POSTFIELDS, "serial=".$token."&pin=".$pin."&MaxFailCount=".$MaxFailCount."&SyncWindow=".$SyncWindow."&OtpLen=".$OtpLen."&CounterWindow=".$CounterWindow."&hashlib=".$hashlib."&timeWindow=".$timeWindow."&timeStep=".$timeStep."&timeShift=".$timeShift."&countAuthSuccessMax=".$countAuthSuccessMax."&countAuthSuccess=".$countAuthSuccess."&countAuth=".$countAuth."&countAuthMax=".$countAuthMax."&validityPeriodStart=".$validityPeriodStart."&validityPeriodEnd=".$validityPeriodEnd."&phone=".$phone);
        curl_setopt($this->ch, CURLOPT_SSL_VERIFYPEER, false);
        $answer = curl_exec($this->ch);
        if (curl_error($this->ch)) return ("cURL request error!");
        return json_decode($answer);
    }

/*show()
        GET /admin/show
        displays the list of the available tokens
        Parameters
            serial – (optional) only this serial will be displayed
            user –(optional) only the tokens of this user will be displayed. If the user does not exist, WibioOtp will search tokens of users, who contain this substring.
            viewrealm – (optional) takes a realm, only the tokens in this realm will be displayed
            tokeninfo_format – (optional) if set to “json”, this will be supplied in embedded JSON otherwise, string format is returned with dates in format DD/MM/YYYY TODO
*/
    public function adminShowAll($viewrealm='', $tokeninfo_format='json')
    {
        curl_setopt($this->ch, CURLOPT_URL, self::SERVER."/admin/show?viewrealm=".$viewrealm."&tokeninfo_format=".$tokeninfo_format);
        curl_setopt($this->ch, CURLOPT_POST, false);
        curl_setopt($this->ch, CURLOPT_SSL_VERIFYPEER, false);
        $answer = curl_exec($this->ch);
        if (curl_error($this->ch)) return ("cURL request error!");
        return json_decode($answer);
    }
    public function adminShowUser($user, $viewrealm='', $tokeninfo_format='json')
    {
        curl_setopt($this->ch, CURLOPT_URL, self::SERVER."/admin/show?user=".$user."&viewrealm=".$viewrealm."&tokeninfo_format=".$tokeninfo_format);
        curl_setopt($this->ch, CURLOPT_POST, false);
        curl_setopt($this->ch, CURLOPT_SSL_VERIFYPEER, false);
        $answer = curl_exec($this->ch);
        if (curl_error($this->ch)) return ("cURL request error!");
        return json_decode($answer);
    }
    public function adminShowToken($token, $viewrealm='', $tokeninfo_format='json')
    {
        curl_setopt($this->ch, CURLOPT_URL, self::SERVER."/admin/show?serial=".$token."&viewrealm=".$viewrealm."&tokeninfo_format=".$tokeninfo_format);
        curl_setopt($this->ch, CURLOPT_POST, false);
        curl_setopt($this->ch, CURLOPT_SSL_VERIFYPEER, false);
        $answer = curl_exec($this->ch);
        if (curl_error($this->ch)) return ("cURL request error!");
        return json_decode($answer);
    }

/*tokenrealm()
        POST /admin/tokenrealm
        set the realms a token belongs to
        Parameters
            serial – (required) serial number of the token
            realms – (required) comma seperated list of realms
*/
    public function adminTokenRealm($token, $realms)
    {
        curl_setopt($this->ch, CURLOPT_URL, self::SERVER."/admin/tokenrealm");
        curl_setopt($this->ch, CURLOPT_POST, true);
        curl_setopt($this->ch, CURLOPT_POSTFIELDS, "serial=".$token."&realms=".strtolower($realms));
        curl_setopt($this->ch, CURLOPT_SSL_VERIFYPEER, false);
        $answer = curl_exec($this->ch);
        if (curl_error($this->ch)) return ("cURL request error!");
        return json_decode($answer);
    }

/*unassign()
    POST /admin/unassign
    unassigns a token from a user. i.e. the binding between the token and the user is removed
    Parameters
        serial – (required) - the serial number / identifier of the token
*/
    public function adminUnassignToken($token)
    {
        curl_setopt($this->ch, CURLOPT_URL, self::SERVER."/admin/unassign");
        curl_setopt($this->ch, CURLOPT_POST, true);
        curl_setopt($this->ch, CURLOPT_POSTFIELDS, "serial=".$token);
        curl_setopt($this->ch, CURLOPT_SSL_VERIFYPEER, false);
        $answer = curl_exec($this->ch);
        if (curl_error($this->ch)) return ("cURL request error!");
        return json_decode($answer);
    }

/*userlist()
        GET /admin/userlist
        lists the user in a realm
        Parameters
            <searchexpr> – will be retrieved from the UserIdResolverClass
            realm – a realm, which is a collection of resolver configurations
            resConf – a destinct resolver configuration
            page – the number of page, which should be retrieved (optional)
            rp – the number of users per page (optional)
*/
    public function adminUserList($searchexpr='', $realm='', $resConf='', $page=1, $rp=100)
    {
        curl_setopt($this->ch, CURLOPT_URL, self::SERVER."/admin/userlist?searchexpr=".$searchexpr."&realm=".self::REALM."&resConf=".$resConf."&page=".$page."&rp=".$rp);
        curl_setopt($this->ch, CURLOPT_POST, false);
        curl_setopt($this->ch, CURLOPT_SSL_VERIFYPEER, false);
        $answer = curl_exec($this->ch);
        if (curl_error($this->ch)) return ("cURL request error!");
        return json_decode($answer);
    }


}
